ledger-core/virtual/object/object.go

```diff
@@ -341,42 +342,50 @@ func (sm *SMObject) stepWaitState(ctx smachine.ExecutionContext) smachine.StateU
 }
 
 func (sm *SMObject) stepGotState(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	state := sm.GetState()
+
 	if sm.PreviousExecutorOrderedPendingCount == 0 {
-		sm.releaseOrderedExecutionPath(ctx)
-		sm.releaseUnorderedExecutionPath(ctx)
-	} else if sm.GetState() != Empty {
-		sm.releaseUnorderedExecutionPath(ctx)
-	}
+		sm.releaseOrderedExecutionPath()
+		sm.releaseUnorderedExecutionPath()
+	} else {
+		var pendingConstructorFinishedCtl smsync.BoolConditionalLink
+		if state == Empty {
+			pendingConstructorFinishedCtl = smsync.NewConditionalBool(false, "pendingConstructorFinished")
+			sm.PendingConstructorFinished = pendingConstructorFinishedCtl.SyncLink()
+		} else {
+			sm.releaseUnorderedExecutionPath()
+		}
 
-	if sm.PreviousExecutorOrderedPendingCount > 0 {
-		sm.SignalPendingsFinished = ctx.NewBargeIn().
-			WithJump(sm.stepReleaseExecutionPaths)
-	}
+		sm.SignalOrderedPendingFinished = ctx.NewBargeInWithParam(func(interface{}) smachine.BargeInCallbackFunc {
+			return func(ctx smachine.BargeInContext) smachine.StateUpdate {
+				if state == Empty {
+					smachine.ApplyAdjustmentAsync(pendingConstructorFinishedCtl.NewValue(true))
+					sm.releaseUnorderedExecutionPath()
```

Comment:
If execution is blocked before tolerance semaphore then we don't need extra blocking of unordered semaphor, as far as I understood

Reply:
`sm.releaseUnorderedExecutionPath()` - only if `state != Empty`. If it's empty then we won't unblock it.

Reply:
Disregard my comment. Unordered is blocked until either constructor is finished in current pulse, or pending constructor result is received from previous executor.